Breaking out ClientCallFactory abstract class

nmittler opened the pull request:
Overview of the changes:

- Renamed Channel to ClientCallFactory
- Created a new Channel interface containing the public methods from ChannelImpl
- Modified ClientInterceptor to be a factory for ClientCallFactory instances.
- Modified the compiler to use ClientCallFactory rather than Channel.

This takes several steps towards addressing #545.

Some of the discussion leading up to this is here: https://github.com/grpc/grpc-java/issues/64#issuecomment-112221220

nmittler commented:
@louiscryan @ejona86 @wrwg @zhangkun83 PTAL.  I suspect that this may be a bit controversial, but I'd like to get the discussion started.

wrwg commented on auth/src/main/java/io/grpc/auth/ClientAuthInterceptor.java:
```diff
@@ -34,8 +34,8 @@
 import com.google.auth.Credentials;
 import com.google.common.base.Preconditions;
 
+import io.grpc.CallFactory;
```
Should it be ClientCallFactory? I think it would need to be for consistency.

The same may apply to unchanged CallOptions?

zhangkun83 replied:
I agree with ClientCallFactory. I don't think it applies to CallOptions. ClientCall and ServerCall are the representations of a logical Call on each side, while CallOptions is the options for the logical Call, not just its client-side representation. For example, one of the options, deadline, applies to both sides.

nmittler replied:
done.

wrwg commented on core/src/main/java/io/grpc/CallFactory.java:
```diff
@@ -0,0 +1,70 @@
+ * A factory for manufacturing {@link ClientCall} instances.
+ */
+@ThreadSafe
+public abstract class CallFactory {
```
ClientCallFactory or ClientCallProvider?

nmittler replied:
done.

wrwg commented on core/src/main/java/io/grpc/Channel.java:
```diff
@@ -1,78 +1,78 @@
+  /**
+   * Returns the factory used for making new calls on this channel.
+   */
+  CallFactory callFactory();
```
I think it's cleaner to have this functionality factored out. +1

wrwg commented on core/src/main/java/io/grpc/ClientInterceptor.java:
```diff
@@ -31,35 +31,74 @@
 @ThreadSafe
-public interface ClientInterceptor {
+public abstract class ClientInterceptor extends CallFactory {
+  private CallFactory next;
```
I feel the design to make the interface a stateful abstract class is a bit concerning. In many cases, an interceptor chain can be created as an immutable object.

Two alternatives:
- keep the old design and drop implementation of CallFactory
- strengthen the contract and introduce getNext as abstract/interface method. Then subclasses decide how to create and store chaining.

nmittler replied:
Agreed that this implementation is a bit hacky :).

Actually the more I think about it, I don't think what we really need is an `InterceptorFactory`, not `Interceptor`. The call to create an interceptor chain should just take a list of these factories which will look like:

```java
interface InterceptorFactory {
    ClientCallFactory newInterceptor(ClientCallFactory next);
}
```

I'll sketch this out and update the PR shortly.

wrwg commented:
This looks generally good to me.

zhangkun83 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -57,7 +57,7 @@
 /** A communication channel for making outgoing RPCs. */
 @ThreadSafe
-public final class ChannelImpl extends Channel {
+public final class ChannelImpl extends CallFactory implements Channel {
```
If `ChannelImpl` needs to extend `CallFactory` too, it's an indication that `Channel` as an interface is not a complete abstraction for its implementations. It would look cleaner if `ChannelImpl` only implemented `Channel`, and you called `ChannelImpl.callFactory()` whenever a `CallFactory` is required.

nmittler replied:
Agreed ... done.

nmittler commented:
@wrwg @zhangkun83 PTAL

zhangkun83 commented on core/src/main/java/io/grpc/ClientInterceptorFactory.java:
```diff
@@ -0,0 +1,43 @@
+  /**
+   * Creates an interceptor that decorates the given next {@link ClientCallFactory} in the chain.
+   */
+  ClientCallFactory newInterceptor(ClientCallFactory next);
```
This interface looks much cleaner than the original, but I have issues with the naming. Given there isn't a class called "Interceptor", having names referring to "creating interceptors" confuses me. Maybe `interface ClientInterceptor { intercept() }`? Or being more straightforward, `interface ClientCallFactoryDecorator { decorate() }`?

nmittler replied:
SGTM, I had the exact same thought :)  .. done.

zhangkun83 commented on core/src/main/java/io/grpc/ChannelImpl.java:
```diff
@@ -57,12 +57,13 @@
   private final ExecutorService executor;
   private final String userAgent;
   private final Object lock = new Object();
+  private final ClientCallFactory callFactory = new InternalCallFactory();
```
Given this is the only instance of InternalCallFactory, I slightly prefer it being an anonymous class.

nmittler replied:
I actually didn't end up needing it.

zhangkun83 commented:
Had two comments. Otherwise LGTM. I like the fact that ClientInterceptor no longer copycats the Channel interface.

nmittler commented:
@wrwg @zhangkun83 I think I've addressed the comments so far.  PTAL.

zhangkun83 commented on stub/src/main/java/io/grpc/stub/AbstractStub.java:
```diff
@@ -49,34 +49,34 @@
    */
-  public Channel getChannel() {
-    return channel;
+  public ClientCallFactory getChannel() {
```
getClientCallFactory()?

nmittler replied:
done.

zhangkun83 commented on stub/src/main/java/io/grpc/stub/AbstractStub.java:
```diff
@@ -113,20 +113,20 @@ public final S withDeadlineNanoTime(@Nullable Long deadlineNanoTime) {
    */
-  public final S withChannel(Channel newChannel) {
-    return build(newChannel, callOptions);
+  public final S withChannel(ClientCallFactory newCallFactory) {
```
withClientCallFactory()?

nmittler replied:
done.

zhangkun83 commented on stub/src/main/java/io/grpc/stub/AbstractStub.java:
```diff
@@ -89,10 +89,10 @@ public CallOptions getCallOptions() {
   /**
    * Returns a new stub with the given channel for the provided method configurations.
```
channel -> call factory.

There are still "channel" throughout this file. They should all be changed to call factory.

nmittler replied:
ah right ... fixed all the occurrences in this file.

nmittler commented:
@zhangkun83 I've migrated nearly all references to ChannelImpl over to Channel. PTAL

nmittler commented:
/cc @carl-mastrangelo you may also be interested in this since you've been working on cleaning up our API.

ejona86 commented:
It doesn't seem very useful for me to review this, given I've already shared my opinions about it on the issues, and I don't see how it will combine with #700 and #28, which seem like they need new methods on this PR's `CallFactory`, which makes it a strange factory.

nmittler commented:
@ejona86 I know that you have very strong opinions about the design decisions made in this PR (as we've discussed offline previously).  As I mentioned above, I expected this to be controversial but I'd like to start a discussion to see if we can come to some sort of resolution.  I'm definitely not alone in expressing concern that parts of our API have become a bit awkward. This is a first attempt toward addressing some of those concerns.  The PR has a few parts and I suspect you're not against all of it.  Perhaps it would be more helpful if you pointed out the parts that are causing you heartburn?

ejona86 commented:
> Renamed Channel to ClientCallFactory

The idea that "channel has nothing specific about a channel" seems true today, but doesn't seem true tomorrow. The apparent reasoning for renaming it to "factory" is "because it has one method that looks like a factory". So when it no longer has one method, does it really make sense? It _will_ be getting more methods.

> Created a new Channel interface containing the public methods from ChannelImpl

I still am against making an interface for the channel, because it seems to serve no point. The only argument for it I've heard was along the lines of "it makes me feel better."

Easily fixable: it is an interface instead of an abstract class. To be able to add any method to it, it must be an abstract class.

> Modified ClientInterceptor to be a factory for ClientCallFactory instances.

Yet another layer of indirection, with no real benefit that I can see, other than giving us a good feeling inside that we were able to reuse an interface. Practically, it doesn't seem like it would aid users and would just be annoying for interceptors (as can be seen by ClientAuthInterceptor).

* Having Channel/ChannelImpl have a method to get a call factory.

I don't really get what is gained vs just "extends ClientCallFactory", in part because I don't agree with having the separate Channel interface, and in part because I don't see how it will make sense when methods are added to ClientCallFactory over time.

* The sum of the parts

This is a lot of API churn. Churn brings risk. Such changes need to have a strong enough need to warrant the risk, and we won't have any stabilization time in order to make small tweaks. Yet I don't see the practical needs being solved by the changes. The only piece that seems to be trying to solve an existing problem is renaming Channel to ClientCallFactory which seeks to provide a more accurate name, but that I think is a short-term optimization ignoring the new methods that Channel will get over time.

nmittler commented:
> The idea that "channel has nothing specific about a channel" seems true today, but doesn't seem true tomorrow. The apparent reasoning for renaming it to "factory" is "because it has one method that looks like a factory". So when it no longer has one method, does it really make sense? It will be getting more methods.

The reasoning behind renaming the current `Channel` interface to `CallFactory` is that the name `Channel` is often used to refer to some sort of connection object. Along with that, a `Channel` api would typically include connection lifecycle management, etc.

Whether `CallFactory` ends up being the final name is perhaps TBD, but I think the separation of call creation from connection management is still a useful distinction in that it clarifies roles for different types of users. The main application may be interested in knowing the state of the connection (e.g. ready to send RPCs, wait until shutdown). This is much different from the role of stubs and interceptors, which just need to be able to create RPCs, and giving them an API that also enables shutdown of the connection would be confusing and could lead to programming error.

Regarding new methods for `CallFactory`, I expect we'll have some discussion to determine the appropriate place for these methods, assuming additional methods are in fact needed.  We can revisit the interface changes on a case-by-case basis and our API will evolve naturally.  I don't see this point as a problem with the design proposed in this PR, rather I believe your concern is simply the name of the interface.  Perhaps `CallFactory` may change, however I still assert that `Channel` is still not the right name here.

> I still am against making an interface for the channel, because it seems to serve no point. The only argument for it I've heard was along the lines of "it makes me feel better." Easily fixable: it is an interface instead of an abstract class. To be able to add any method to it, it must be an abstract class.

+1, I'm fine with making it an abstract class ... in fact, I was already considering doing that. I believe having an abstract `Channel` that contains the full interface (including channel lifecycle management, etc) is useful. While there may never be another channel, there is no reason that I've heard to force this restriction on the end-user. If they want to create their own `Channel` or even just create a mock for testing, they should be able to do so without us getting in the way. In addition, forcing a single `Channel` goes against API stability in that if we ever find a need for a second channel implementation, we simply have to break the API in order to refactor.

> Yet another layer of indirection, with no real benefit that I can see, other than giving us a good feeling inside that we were able to reuse an interface. Practically, it doesn't seem like it would aid users and would just be annoying for interceptors (as can be seen by ClientAuthInterceptor).

The current `ClientInterceptor` and `Channel` APIs are basically a copy & paste with the exception that `ClientInterceptor` also take a `next` object. This redundant interface is confusing since it establishes two independent classes of objects that are responsible for creating `ClientCall`s in gRPC. Unifying them to use a single interface is a fairly significant simplification of the API. It also simplifies the maintenance of the API going forward as the requirements for call creation evolve over time (e.g. we just added `CallOptions`).

> This is a lot of API churn. Churn brings risk. Such changes need to have a strong enough need to warrant the risk, and we won't have any stabilization time in order to make small tweaks. Yet I don't see the practical needs being solved by the changes. The only piece that seems to be trying to solve an existing problem is renaming Channel to ClientCallFactory which seeks to provide a more accurate name, but that I think is a short-term optimization ignoring the new methods that Channel will get over time.

Agreed that this is churn, but I would argue that it's done in the spirit of making the API more understandable and maintainable going forward. I don't believe that this PR brings with it a lot of risk as it's not changing any logic ... it's just restructuring some parts of our API. Yes, we may need to expand our API in the future but it's not clear that this change makes future refactoring any more difficult than the current API, and I would argue that this PR will actually make it easier.

ejona86 commented:
> The main application may be interested in knowing the state of the connection (e.g. ready to send RPCs, wait until shutdown). This is much different from the role of stubs and interceptors, which just need to be able to create RPCs

So this is a point we disagree, and I think is a focal point of much of the disagreement.

> Regarding new methods for CallFactory, I expect we'll have some discussion to determine the appropriate place for these methods, assuming additional methods are in fact needed.

Yes. I think that discussion should happen before these changes, but maybe in light of these changes.

> We can revisit the interface changes on a case-by-case basis and our API will evolve naturally.

Except that I'm arguing that the current API is better for the types of evolution we're expecting.

> I don't see this point as a problem with the design proposed in this PR, rather I believe your concern is simply the name of the interface. Perhaps CallFactory may change, however I still assert that Channel is still not the right name here.

I think we disagree on what Channel _is_, and thus disagree on what it's name should be and what methods will be on it.

> While there may never be another channel, there is no reason that I've heard to force this restriction on the end-user.

We could make an interface for each method we have defined anywhere. However, who does that help? We shouldn't make new interfaces/classes based on what we _can_ do, but what is helpful. If there is no argument that it is helpful, or the argument is weak, then we favor not having it.

From an API perspective, we have more freedom to change a final concrete class. Changing an abstract class is _more possible_ than an interface, but still more restrictive than a final concrete class. Adding an additional overload to a method gets icky with an abstract class, but is straight-forward with a final concrete class.

I know we disagree on this point, but I feel it also helps in comprehension or our API. If I see the abstract class, I would wonder "where is another implementation?" and when I find there isn't one I think "when/why would there be another implementation?" The fact that we can't answer these questions proves to me that it shouldn't exist.

> If they want to create their own Channel or even just create a mock for testing, they should be able to do so without us getting in the way.

Why do they want to create their own channel, and be required to use the same lifecycle, etc? They can create their own CallFactory. What is the use case here?

Mocks suck, or rather they suck when overused/when used in inappropriate cases. Mocking Channel is not appropriate. You need a fake or something similar instead.

If we need an abstract Channel in order to develop a fake, then that could be fine. But that's not been the argument and we should have that discussion when we are making the fake.

> In addition, forcing a single Channel goes against API stability in that if we ever find a need for a second channel implementation, we simply have to break the API in order to refactor.

Why is that? `ChannelImpl` is _never_ passed to our API. We only care about `Channel`. We could add it later without breaking API.

Except, not really.

Our API is fundamentally broken if there is more than one `Channel` implementation (outside of testing), even with this PR: you can only create transports with `ChannelImpl`. We would need to fix `AbstractChannelBuilder`, which has been proposed. But the point is that this is _not_ a localized/small/trivial change; we would need to re-think our construction patterns, and I don't think we could do that quickly.

> The current ClientInterceptor and Channel APIs are basically a copy & paste with the exception that ClientInterceptor also take a next object. This redundant interface is confusing since it establishes two independent classes of objects that are responsible for creating ClientCalls in gRPC.

The _current_ API. Channel will get methods that ClientInterceptors can't override. But even so, I claim it isn't confusing because an object is _clearly_ one or the other. They serve distinctly different purposes. I would agree that it is _unfortunate_, but I see the solution as worse than the problem. And practically I don't think it would cause problems.

And really, only one creates wholly new RPCs: Channel. Interceptor should always be delegating to a channel.

> Unifying them to use a single interface is a fairly significant simplification of the API.

Really? Significant? Simplification? I don't see it. I see the same number of concepts (Copy-pasted Channel vs CallFactoryFactory) and the same number of methods. Applications using interceptors would be unchanged. I see an extra level of indirection for Interceptor implementors, and no way to enforce/discourage interceptors from changing the behavior of some methods (say, channel connectivity). The extra layer of indirection is probably what bothers me the most; the number of layers of indirection for an interceptor was already more than I would have preferred (but, there doesn't seem to be a way to remove the existing layers).

> It also simplifies the maintenance of the API going forward as the requirements for call creation evolve over time (e.g. we just added CallOptions).

I find that to be a feature. We get to chose which things are reasonable for an interceptor to override.

In any case, "simplifies maintenance" seems out of proportion. The code supporting interceptors is _trivial_. Yes, it may simplify maintenance, but it is so negliable... In the case of adding CallOptions, the "simplied maintenance" would have been a _single_ line of code difference, not needing to add a single parameter to a single a method declaration.

> I don't believe that this PR brings with it a lot of risk as it's not changing any logic ... it's just restructuring some parts of our API.

The risk is that we don't notice the problems with the API. Risks due to bugs are _much_ less of a concern compared to the risk of issues with an API, as bugs can be fixed but APIs commonly can't.

After the creation of an API, we really need to use it for long enough to have enough experience to notice its downfalls. Even if this PR fixed _all_ the problems with the current API, it would still be risky. The new API could have problems that we wouldn't be aware of until it is too late.

nmittler commented:
I think the list of specific points has grown too long to bother addressing each individually ... let me step back a bit.

It seems that the overarching point of disagreement is over the concept of what a `Channel` is. Currently, it is:

1. An abstraction over the connectivity to the server (i.e. `ChannelImpl`). This part of the interface is currently only used by the application.
2. A facility for creating calls. This is used by the application if using the `Channel` API directly, or by the generated stubs.
3. An interceptor of call creation that can change the behavior of the calls (e.g. adding metadata).

My definition for an interceptor in bullet 3 has been the definition that we've been going with up to this point, but it sounds like you're suggesting that it should be also able to intercept all capabilities of the channel, such as connectivity state or termination.  Does that accurately capture you're view of how the role of interceptors is expanding?

Regardless, I do think that bullet 2 warrants a more restrictive interface, even if it is just a couple of methods for creating calls.  This is the main use case for stubs and is conceptually orthogonal from connectivity or connection state.  With that said, this is the only separate interface I'm proposing. Nowhere have I suggested creating an interface per method; only that we break out parts of the API that go together and can be useful independently into separate interfaces.

ejona86 commented:
> My definition for an interceptor in bullet 3 has been the definition that we've been going with up to this point, but it sounds like you're suggesting that it should be also able to intercept all capabilities of the channel, such as connectivity state or termination.

No. I am arguing that it _shouldn't_.

Today, bullet 3 isn't really true. `ClientInterceptor` is a completely different interface from `Channel`. They are similar and related, and we have adapters, but they are distinct.

We are going to need more methods on `Channel`, and those are going to need to be accessed by interceptors and the application. Connectivity state is one example, but debatable. A non-debatable example is the authority()-like string necessary for JWT.

However, by conflating interceptors and channels by making them the same interface, we now can't extend one interface separately from the other.

> Does that accurately capture you're view of how the role of interceptors is expanding?

That states my view of how _this PR_ expands the role of interceptors.

> Regardless, I do think that bullet 2 warrants a more restrictive interface, even if it is just a couple of methods for creating calls.

That is `Channel` today, before this PR. No changes necessary.

However, I will note that `Channel` _needs_ to expand; a `CallFactory` (with one method) is _not_ a powerful enough interface. `Channel` should still stay minimal, but it _will_ be more than just a factory.

> With that said, this is the only separate interface I'm proposing. Nowhere have I suggested creating an interface per method

I only brought up interface-per-method to help illustrate why I feel the argument was a fallacy.

nmittler commented:
> a CallFactory (with one method) is not a powerful enough interface.

Here I *strongly* disagree. I find `CallFactory` (or whatever you want to call it) to be a core interface of gRPC and fulfills a common need across application, interceptor and stub. @wrwg had also come to a similar conclusion when defining his `Callable` API.  It's a useful abstraction, regardless of how many methods it contains (currently 2).

>> Does that accurately capture you're view of how the role of interceptors is expanding?

> That states my view of how this PR expands the role of interceptors.

This PR does in no way expand the role of interceptors. They have the same role as before, they just make use of the `CallFactory` interface.

> We are going to need more methods on Channel, and those are going to need to be accessed by interceptors and the application. Connectivity state is one example, but debatable. A non-debatable example is the authority()-like string necessary for JWT.
However, by conflating interceptors and channels by making them the same interface, we now can't extend one interface separately from the other.

I think it's fair to say that *most* interceptors will not need to access parts of the channel beyond call creation.  The auth interceptor only needs the authority.  Rather than providing the full channel API to all interceptors, I suspect a better solution would be to treat these as one-off cases and to restrict access of the channel to only the interceptors that explicitly need it.  In the Spring Framework, for example, an object that depends on the application context is free to implement [ApplicationContextAware](http://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/context/ApplicationContextAware.html).  I don't think having an interceptor that implements `ChannelAware` would be such a bad approach.

nmittler commented:
@louiscryan @wrwg it would be good to hear your thoughts on this.

ejona86 commented:
> I suspect a better solution would be to treat these as one-off cases and to restrict access of the channel to only the interceptors that explicitly need it. In the Spring Framework, for example, an object that depends on the application context is free to implement ApplicationContextAware. I don't think having an interceptor that implements ChannelAware would be such a bad approach.

That seems like a very different design, because it doesn't involve wrapping/delegation at all. I don't quite see how it would work for us.

Could we may this more concrete? Today, `ClientInterceptors` has:
```java
public Channel intercept(Channel channel, ClientInterceptor... interceptors);
```

What would that change to? Stubs currently allow reconfiguration by adding new interceptors (and it _must_ support auth). How would that work if they only have a `CallFactory`?

nmittler commented:
>>  I suspect a better solution would be to treat these as one-off cases and to restrict access of the channel to only the interceptors that explicitly need it. In the Spring Framework, for example, an object that depends on the application context is free to implement ApplicationContextAware. I don't think having an interceptor that implements ChannelAware would be such a bad approach.

Let me illustrate this idea a little bit so we have something more concrete to discuss.  Suppose we have an interface like this:

```java
interface ChannelAware {
    void setChannel(Channel channel);
}
```

The builder for the channel is given a list of interceptors from which it will build the `CallFactory` used by the channel.  The contract of the builder will be such that after the channel is fully built (i.e. all interceptors have been created), it takes another pass through the interceptors as call `setChannel` on any interceptors implementing `ChannelAware`. The reason we can't pass the `Channel` to `ClientInterceptors` is because in this case the channel hasn't been built yet.

ejona86 commented:
As discussed offline, that doesn't support per-RPC auth credentials, which is a required feature. Per-RPC auth would be very common for 3LO.

nmittler commented:
@ejona86 I see what you mean.  Specifically, the problem is this method from `AbstractStub`:

```java
public final S withInterceptors(ClientInterceptor... interceptors) {
    return build(ClientInterceptors.intercept(channel, interceptors), callOptions);
  }
```

It actually seems a little strange that the stubs *know* about interceptors at all ... I would expect they wouldn't.  Perhaps a simple solution to this would be to just allow building with a replacement `CallFactory`.

```java
public final S withCallFactory(CallFactory cf) {
    return build(cf, callOptions);
  }
```

`AbstractStub` already provides`getCallFactory`, so intercepting it would be fairly straight forward. I think overall, this is a more flexible interface as we're not tying the stub to a `CallFactory` indefinitely.

ejona86 commented:
@nmittler, that seems to simply avoid acknowledging the problem. If you were provided a stub, you still can't reconfigure the stub per-call because you need a `Channel` and not a `CallFactory`.

The reason it is on the stub is because it will be a common reconfiguration use case. Probably only second to setting a deadline.

What is the hoped-for benefit of using `CallFactory` instead of `Channel` in the stub?

nmittler commented:
> The reason it is on the stub is because it will be a common reconfiguration use case. Probably only second to setting a deadline.

@ejona86 Now that the build process of the channel incorporates addition of interceptors, I'm not sure that stub reconfiguration with interceptors will be as common a use case.

For the remaining users that absolutely want to reconfigure the stub with interceptors, I don't think that asking them to call `ClientInterceptors.intercept` is unreasonable.

> What is the hoped-for benefit of using CallFactory instead of Channel in the stub?

Generally, it's a separation of concerns.  The stub only needs to be able to create calls, it doesn't need (nor should it have the ability) to modify channel state.

ejona86 commented:
> For the remaining users that absolutely want to reconfigure the stub with interceptors, I don't think that asking them to call ClientInterceptors.intercept is unreasonable.

I was saying they _can't_ call `ClientInterceptors.intercept`, because they won't have a reference to `Channel`.

> The stub only needs to be able to create calls

Sorta, because of the above.

> it doesn't need (nor should it have the ability) to modify channel state.

I _completely_ agree. And _that_ is why Channel today doesn't have shutdown.

nmittler commented:
@louiscryan @wrwg it would be good to hear your thoughts on this topic.

nmittler commented:
Based on offline discussion, this PR is abandoned.  Will follow up with other PRs to address the class hierarchy.
